base/src/main/java/com/smartdevicelink/session/BaseSdlSession.java

```diff
@@ -187,19 +190,57 @@ public void setSdlSecurity(SdlSecurityBase sec) {
     protected void processControlService(ProtocolMessage msg) {
         if (sdlSecurity == null)
             return;
+
+        SecurityQueryPayload receivedHeader = SecurityQueryPayload.parseBinaryQueryHeader(msg.getData().clone());
+        if (receivedHeader == null) {
+            DebugTool.logError(TAG, "Malformed Security Query Header");
+            return;
+        }
+
         int iLen = msg.getData().length - 12;
         byte[] data = new byte[iLen];
         System.arraycopy(msg.getData(), 12, data, 0, iLen);
 
         byte[] dataToRead = new byte[4096];
 
-        Integer iNumBytes = sdlSecurity.runHandshake(data, dataToRead);
+        Integer iNumBytes = null;
+
+        if (receivedHeader.getQueryID() == QueryID.SEND_INTERNAL_ERROR
+                && receivedHeader.getQueryType() == QueryType.NOTIFICATION) {
+            if (receivedHeader.getErrorCode() != null) {
+                DebugTool.logError(TAG, "Client internal error: " + receivedHeader.getErrorCode().getName());
+            } else {
+                DebugTool.logError(TAG, "Client internal error: No information provided");
+            }
+            return;
+        }
 
-        if (iNumBytes == null || iNumBytes <= 0)
+        if (receivedHeader.getQueryID() != QueryID.SEND_HANDSHAKE_DATA
+                && (receivedHeader.getQueryType() != QueryType.NOTIFICATION || receivedHeader.getQueryType() != QueryType.REQUEST)) {
```

Comment:
`receivedHeader.getQueryType() != QueryType.NOTIFICATION || receivedHeader.getQueryType() != QueryType.REQUEST` is always true, because the queryType cannot be both `NOTIFICATION` and `REQUEST` at the same time. One of these expressions will always be `true`.